keps/sig-auth/3926-handling-undecryptable-resources/README.md

```diff
@@ -0,0 +1,955 @@
+  //        - "type" CauseType - CauseTypeUnexpectedServerResponse
+  //        - "message" string - the error message from the store backend
+  //        - "field" string - the full path with the key of the resource that failed reading
+  //
+  // Status code 500
+  StatusReasonStoreReadError StatusReason = "StorageReadError"
+```
+
+This error will also include full paths to the resources that cannot be read in
+the unstructured, human-readable message.
+
+With this error, a new `ListOption` can be introduced - `IgnoreStorageReadError`.
+
+```go
+type ListOptions struct {
+  ...
+  // IgnoreStorageReadError ignores errors that might occur during reading a
+  // resource from the storage backend. Instead, the message of that error gets
+  // transformed into a warning.
+  // +optional
+  IgnoreStorageReadError bool
+}
+```
+
+This option allows the caller to ignore the resources that failed transformation
+and still retrieve the list of resources that remain readable.
+
+
+#### Deleting a failing resource
+
+Deleting a resource is a rather complicated process:
+1. a resource might represent an actual process running on a host (Pod)
+2. there might be other resources with owner references to the resource that's being deleted
+3. a resource might contain finalizers that safeguard the deletion of the given resource
+   before other dependent resources are deleted (typically - namespaces and the `kubernetes` finalizer)
+
+An unconditional deletion should try to do best effort on all of the above, but
+in case of an undecryptable resource, all the above would be ignored.
+
+For case 1., ignoring an underlying process may not be an issue as kubelet is supposed
+to take care [of unused containers](https://kubernetes.io/docs/concepts/architecture/garbage-collection/#containers-images).
+
+In case 2., there might be issues with setting related objects as `orphans`, which
+could potentially cause an unwanted cascade deletion of objects.
+
+3. has a potential of becoming rather serious. Finalizers are typically set to
+safeguard other objects, and so if e.g. an aggregated API server is removed, its
+API objects might be scattered around the etcd database without and API to remove them.
+
+To allow unconditional deletion, a new `DeleteOption` should be introduced - `UnconditionalDeleteWithClusterBreakingPotential`
+
+```go
+type DeleteOptions struct {
+  ...
+  // UnconditionalDeleteWithClusterBreakingPotential will still try to perform the normal
+  // deletion flow but if any errors occur, it will disregard them and perform deletion
+  // of the given resource nonetheless.
+  // WARNING: This will break the cluster if the resource has dependencies beyond
+  //          the caller's comprehension. Use only if you REALLY know what you are
+  //          doing.
+  // WARNING: Vendors will most likely consider using this option to be breaking the
+  //          support of their product.
+  UnconditionalDeleteWithClusterBreakingPotential bool
```

Comment:
it's unclear what this API lets you skip... I would not expect to be allowed to skip admission or finalizers if the existing persisted object can be decoded successfully

Reply:
+1 - I thikn it should only allow deleting objects (with all the consequences) that we cannot decode. Nothing else should be allowed.

Reply:
Just to get our lingo synchronized - I would consider `decoding` and `transforming` two separate cases (you'll see why in code ref below).

What about objects we cannot retrieve from storage for other reasons?

The generic storage Get implementation fails in one additional case, that is if key cannot be constructed - https://github.com/kubernetes/kubernetes/blob/3cf9f66e90d560ac080687610933c712bcf37b39/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L756-L769

The etcd3 store implementation fails in 6 other cases:
https://github.com/kubernetes/kubernetes/blob/3cf9f66e90d560ac080687610933c712bcf37b39/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L132-L166

Most of the cases indeed seem like errors we may not want to ignore. Should we focus solely on the transformation error, then?

Reply:
> Should we focus solely on the transformation error, then?

I don't think so... transformation and decoding failures have almost identical implications, symptoms, and considerations for deleting anyway. Solving both problems together and consistently makes sense to me